src/dbnode/storage/index/wide_query_results.go

```diff
@@ -0,0 +1,231 @@
+		},
+		batchCh: opts.IndexBatchCollector,
+		shards:  opts.ShardsQueried,
+	}
+
+	if len(opts.ShardsQueried) > 0 {
+		// Only apply filter if there are shards to filter against.
+		results.shardFilter = shardFilter
+	}
+
+	return results
+}
+
+func (r *wideResults) AddDocuments(batch []doc.Document) (int, int, error) {
+	if r.closed || r.pastLastShard {
+		return 0, 0, ErrWideQueryResultsExhausted
+	}
+
+	err := r.addDocumentsBatchWithLock(batch)
+	release := len(r.batch.IDs) == r.batchSize
+	if release {
+		r.releaseAndWait()
+		r.releaseOverflow(false)
+	}
+
+	return 0, 0, err
+}
+
+func (r *wideResults) releaseOverflow(forceRelease bool) {
+	var (
+		incomplete bool
+		size       int
+		overflow   int
+	)
+	for {
+		size = r.batchSize
+		overflow = len(r.idsOverflow)
+		if overflow == 0 {
+			// NB: no overflow elements.
+			return
+		}
+
+		if overflow < size {
+			size = overflow
+			incomplete = true
+		}
+
+		r.batch.IDs = append(r.batch.IDs, r.idsOverflow[0:size]...)
+		r.batch.IDs = r.batch.IDs[:size]
+		copy(r.idsOverflow, r.idsOverflow[size:])
+		r.idsOverflow = r.idsOverflow[:overflow-size]
+		if !forceRelease && incomplete {
+			return
+		}
+
+		r.releaseAndWait()
+	}
+}
+
+func (r *wideResults) addDocumentsBatchWithLock(batch []doc.Document) error {
+	for i := range batch {
+		err := r.addDocumentWithLock(batch[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *wideResults) addDocumentWithLock(d doc.Document) error {
+	if len(d.ID) == 0 {
+		return errUnableToAddResultMissingID
+	}
+
+	var tsID ident.ID = ident.BytesID(d.ID)
+
+	// Need to apply filter if set first.
+	if r.shardFilter != nil {
+		filteringShard := r.shards[r.shardIdx]
+		shard, shardOwned := r.shardFilter(tsID)
+		// NB: Check to see if shard is exceeded first (to short circuit earlier if
+		// the current shard is not owned by this node, but shard exceeds filtered).
+		if filteringShard > shard {
+			// this element is from a shard lower than the next shard allowed.
+			return nil
+		}
+
+		for filteringShard < shard {
+			// this element is from a shard higher than the next shard allowed;
+			// advance to the next shard, then try again.
+			r.shardIdx = r.shardIdx + 1
+			if r.shardIdx >= len(r.shards) {
+				// shard is past the final shard allowed by filter, no more results
+				// will be accepted.
+				r.pastLastShard = true
+				return ErrWideQueryResultsExhausted
+			}
+
+			filteringShard = r.shards[r.shardIdx]
+			if filteringShard > shard {
+				// this element is from a shard lower than the next shard allowed.
+				return nil
+			}
+		}
+
+		if !shardOwned {
+			return nil
+		}
+	}
+
+	// Pool IDs after filter is passed.
+	tsID = r.idPool.Clone(tsID)
+	if len(r.batch.IDs) < r.batchSize {
+		r.batch.IDs = append(r.batch.IDs, tsID)
+	} else {
+		r.idsOverflow = append(r.idsOverflow, tsID)
+	}
+
+	return nil
+}
+
+func (r *wideResults) Namespace() ident.ID {
+	return r.nsID
+}
+
+func (r *wideResults) Size() int {
+	return 0
```

Comment:
Probably need to update ya?

Reply:
I'll add a comment as to why these return 0s, for avoiding hitting limits... although now that I think about it that may not be sufficient if a different query had already tripped those limits; may refactor the limit detection a bit and make it clearer why it's fine for this to no-op

Reply:
Went ahead with the refactor here 👍